index.js

```diff
@@ -384,9 +391,53 @@ class TrezorKeyring extends EventEmitter {
     });
   }
 
-  signTypedData() {
-    // Waiting on trezor to enable this
-    return Promise.reject(new Error('Not supported on this device'));
+  /**
+   * EIP-712 Sign Typed Data
+   */
+  async signTypedData(address, data, { version }) {
+    const dataWithHashes = transformTypedData(data, version === 'V4');
+
+    // set default values for signTypedData
+    // Trezor is stricter than @metamask/eth-sig-util in what it accepts
+    const {
+      types: { EIP712Domain = [], ...otherTypes } = {},
+      message = {},
+      domain = {},
+      primaryType,
+      // snake_case since Trezor uses Protobuf naming conventions here
+      domain_separator_hash, // eslint-disable-line camelcase
+      message_hash, // eslint-disable-line camelcase
+    } = dataWithHashes;
+
+    // This is necessary to avoid popup collision
+    // between the unlock & sign trezor popups
+    const status = await this.unlock();
+    await wait(status === 'just unlocked' ? DELAY_BETWEEN_POPUPS : 0);
+
+    const response = await TrezorConnect.ethereumSignTypedData({
```

Comment:
I'm still not sure if the params are correct here. I think you either need `path`, `data`, and `metamask_v4_compat` OR `path`, `domain_separator_hash`, and `message_hash`, not all of these together.

Since it seems to be working, it must be ignoring the extra properties (since the expected object shape matches), but it may be better to make the params conditional on Model T vs Model 1.

Reply:
you're getting your protobuf messages and Connect calls confused.

Trezor Connect exposes a single call, that has mandatory `data` and optional `message_hash` and `domain_separator_hash`. Internally it picks whichever is appropriate for the current Trezor and sends that.

Requiring the caller (eth-trezor-keyring) to do it would needlessly complicate code for everyone.

Reply:
Fair enough. Please ignore my earlier comment then.

Reply:
I believe the `trezor-connect` types are out-of-date.

It should be:
  - `path`, `data`, `metamask_v4_compat` = Trezor T support.
  - `path`, `domain_separator_hash`, `message_hash` = Trezor One support.
  - `path`, `data`, `metamask_v4_compat`, `domain_separator_hash`, `message_hash` = Both Trezor One and Trezor T support (what we're using in this PR).

I'm not an expert in TypeScript, but I guess it should be documented something like:

```typescript
function ethereumSignTypedData<T extends Ethereum.EthereumSignTypedDataTypes>(
  params: P.CommonParams & (
    Ethereum.EthereumSignTypedData<T>
    | Ethereum.EthereumSignTypedHash
    | (Ethereum.EthereumSignTypedData<T> & Ethereum.EthereumSignTypedHash)
  )
): P.Response<Protobuf.EthereumTypedDataSignature>;
```

I'll make a PR there. I think the docs also need to be improved too, since they are currently lacking anything about T1 support.